device/devicegate/filter.go

```diff
@@ -0,0 +1,263 @@
+package devicegate
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/xmidt-org/webpa-common/device"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/xhttp"
+)
+
+const (
+	metadataMapLocation = "metadata_map"
+	claimsLocation      = "claims"
+)
+
+var (
+	emptyVal = struct{}{}
+)
+
+type FilterGateOption func(*FilterGate)
+
+type FilterGate struct {
+	filters        map[string]map[interface{}]struct{}
+	allowedFilters map[string]struct{}
+	lock           sync.RWMutex
+}
+
+type filterRequest struct {
+	Key    string
+	Values []interface{}
+}
+
+func NewFilterGate(options ...FilterGateOption) *FilterGate {
```

Comment:
You can dramatically simplify this code if you avoid the use of a constructor and supply and interface that clients uses the `FilterGate` as.  That separates creation time from run time.  For example:

```
type Gate interface {
    // whatever methods you want to expose to clients here:
    Filters() map[string]interface{}
}

type FilterGate struct {
    // export things to simplify construction
    Filters map[string]map[interface{}]struct{}

    //etc
}

// make *FilterGate implement Gate ...

// clients can construct a FilterGate directly, but use it as a Gate:
func NewGate() Gate {
    return &FilterGate{
        Filter: ...
    }
}
```